### ISSUE TITLE
Is an equivalent of `typeguard.check_type` available?

### ISSUE THREAD

**Issue by harshita-gupta** [beartype/beartype#87]
I'd like to use Beartype's custom PEP-compliant validators, but my use case requires that I make an explicit call to a function similar to `typeguard.check_type()`. It doesn't seem like beartype is currently designed to expose a method like this; the decorators are all I can find. How hard would it be to expose a method like this? If I'm given some general guidance I'm happy to attempt a PR for it myself.

**Comment by leycec** [comment]
> I'd like to use Beartype's custom PEP-compliant validators...

That's wonderful! We validate, so you don't have to.

> ...my use case requires that I make an explicit call to a function similar to `typeguard.check_type()`.

That's also wonderful! Well, *sorta.* An exhaustive (and exhausting) discussion currently resides at issue #87, which this issue is *technically* a duplicate of. But your delightful enthusiasm is infectious – so here's the **tl;dr** repurposed specifically for you. Let us help you help us help you recursively until GitHub stackoverflows all over the place.

## The Tl;dr.

An officially supported, highly performant analogue to `typeguard.check_type()` named `beartype.is_bearable()` <sup>*thrilling puns or it didn't happen*</sup> is scoped out for our next *next* `beartype 0.11.0` release.

Our immediately next `beartype 0.10.0` release is expected to drop on Monday, January 2022 – or I don't live in a cabin in the Canadian wilderness with a beautiful wife and two smelly cats. Thankfully, I do.

Ultimately, this means you'd really rather not wait for `beartype 0.11.0` to drop. Instead, @qiujiangkun concocted this [excruciatingly clever drop-in placeholder for `beartype.is_bearable()` that should temporarily satisfy your use case](https://github.com/beartype/beartype/issues/79#issue-1091672434) – unless you deeply care about performance, in which case it shouldn't:

``` python
# If the official is_bearable() tester is available, prefer that.
try:
    from beartype.abby import is_bearable
# Else, fallback to implementing is_bearable() by hand. *audible gulp*
except ImportError:
    def is_bearable(obj, annotation):
        from beartype import beartype
        from beartype.roar import BeartypeCallHintPepReturnException

        @beartype
        def _is_bearable_slow(o) -> annotation:
            return o
    
        try:
            _is_bearable_slow(obj)
            return True
        except BeartypeCallHintPepReturnException as exc:
            pass

        return False
```

## Doesn't Look So Bad If I Squint at It

In theory, that should suffice... *slowly.* In practice, I've tested nothing. Moreover, there are multiple intersecting sources of slowness in the above implementation. Each `is_bearable()` call:

* Declares a largely useless private closure `_is_bearable_slow()` that's just immediately discarded. **Ugh.**
* Invokes the `@beartype` decorator. The type-checking wrapper functions that `@beartype` dynamically generates are *stupidly* fast. And I should know! My wife and cats insinuate I'm stupid on a daily basis. Sadly, this does not extend to the `@beartype` decorator itself – which is *stupidly* slow. Again, I know these things.
* Incurs an additional stack frame by calling `_is_bearable_slow()`. Python function call overhead is notoriously (and surprisingly) high, so this certainly doesn't help matters.
* Coerces raised exceptions into returned booleans. On the one hand, that's actually fast when the passed object satisfies the passed type hint. On the other hand, that's miserably slow when the passed object violates the passed type hint. Exception handling in Python is reasonably fast up until the point at which an exception is raised – *at which point nothing is slower than exception handling in Python.*

## I'm Starting to Squint at That

**Okay.** Let's talk options:

1. **(Option 1)** You're already exhausted and just want this to work. In this case, it'd be super-swell if you'd consider submitting a PR that:
   * Defines a new `beartype.abby` subpackage containing:
     * A trivial `__init__` submodule defining the above `is_bearable()` placeholder. It ain't great, but it works – and that's great.
2. **(Option 2)** You're on the threshold of exhaustion but not quite there. In this case, it'd be super-hyper-swell if you'd consider submitting a PR that:
   * Defines a new `beartype.abby` subpackage containing:
     * A *less* trivial `__init__` submodule memoizing the above `is_bearable()` placeholder. I incoherently discuss this [in detail in an internal `beartype` `#FIXME:` comment found here](https://github.com/beartype/beartype/blob/9abc8c0c4f2faba484365401d15cd6cc996515ed/beartype/__init__.py#L20). The core conceit here is that we can ameliorate the first two (but *not* last two) efficiency concerns above with (*...wait for it*) **caching.** Specifically, we memoize each of the `_is_bearable_slow()` closures generated by each call to `is_beartype()` via a trivial non-thread-safe global dictionary `_is_beartype_hint_to_tester: Dict[object, Callable[[], bool] = {}` mapping from each type hint passed to `is_beartype()` to the corresponding `_is_bearable_slow()` closure internally type-checking the passed object against that type hint. Noice, right? *Right.*
3. **(Option 3)** You're supercharged with victory, vim, vinegar, vitality, and other peppy words prefixed by "v." In this case, it'd be super-hyper-mega-swell if you'd consider submitting a PR that solved *all* of the above efficiency concerns – except the function call overhead. That's unavoidable, 'cause Python gonna Python. Listen, though. You don't want to do this, Harshita. As your free-as-in-beer `beartype` volunteer PR counsel, I advise against doing anything resembling this. In fact, I advise reading any further. Nonetheless, I also encourage this. So, you've read this far against expert adivec. The True Solution looks like this. In the `is_bearable()` tester:
   1. Dynamically generate a new `_is_bearable_fast()` non-closure function whose body is synthesized by calling the ultra-private hidden God-tier function called `beartype._decor._code._pep._pephint.pep_code_check_hint(hint)`. Look. I counselled you against this, didn't I? I assume no responsibility. But... if you call this thing, this thing will return a boolean Python expression evaluating to either ``True`` if an arbitrary object satisfies the passed type hint `hint` *or* ``False`` otherwise. That's exactly what we want, right? *Right.*
   2. Memoize that `_is_bearable_fast()` tester via a trivial non-thread-safe global dictionary `_is_beartype_hint_to_tester: Dict[object, Callable[[], bool] = {}` mapping from each type hint passed to `is_beartype()` to the corresponding `_is_bearable_fast()` tester internally type-checking the passed object against that type hint.
   3. **Rejoice!** For you have made @qiujiangkun's, @rsokl's, and my day and spared my last few working neurons from an ugly debacle that would probably result in a cat scratching me repeatedly.

*Aaaaaand...* that was more information than either of us needed.

**Comment by rsokl** [comment]
> An exhaustive (and exhausting) discussion currently resides at issue #87

Actually it is #79 😄

**Comment by harshita-gupta** [comment]
wow, this is such a wonderfully thorough and detailed response, @leycec. Thank you kindly. I'm going to: 
1. use the code snippet you sent me in the thing I'm currently working on, which I'm currently somewhat hyperfocused on and would like to finish ASAP
2. Return to this thread next week, at which point I hopefully will not be hyperfocused on thing in #1; I will then assess my energy levels and attempt a PR using one of your three very detailed suggestions. 

Thank you again!

**Comment by leycec** [comment]
Thank *you,* Harshita. You've now lit a fire inside of me, too! Option 1 is trivial, because it's there and might even work as intended. I hope to make at least *that* happen sometime over the next week or two – unless someone with more energy than I've enjoyed across an entire lifetime beats me to the punch first. This might even be you. :wink:

Oh... and if that spicy thing you are hyperfocused on is or is going public, please ping me with @leycec when that lands. I'd love to see that secret sauce you've got cooking over there. :cook:

**Comment by leycec** [comment]
*marked as duplicate of #79*

**Comment by harshita-gupta** [comment]
conversations with you are a lot of fun, @leycec. Unfortunately the spicy object of my current hyperfocus is not public, and is unlikely to ever be. But your endless enthusiasm is appreciated as always. 😁

A follow-up question for you on forward compatibility with `is_bearable` – do you anticipate that `is_bearable` will return a `bool` but never throw an exception, as in the implementation [here](https://github.com/beartype/beartype/issues/87#issuecomment-1020856517), or do you imagine that it will return `True` or throw one of `BeartypeCallHintPepReturnException` or `TypeException` (or something else?) as in the implementation [here](https://github.com/beartype/beartype/issues/79#issuecomment-1003521430)?

**Comment by leycec** [comment]
> Unfortunately, the spicy object of my current hyperfocus is not public...

**Drat it all.** Many fantastical things in life are locked behind an unmarked door that will never be accessible to me – and you tell me you just made another. My arch-nemesis: *it's that door.* :door: 

> A follow-up question for you on forward compatibility with `is_bearable`...

Excellent question. Whatever's behind that door, I bet it reeks of awesome, doesn't it? Don't answer that. And your answer is... :drum: 

> ...do you anticipate that `is_bearable` will return a `bool` but never throw an exception, as in the implementation here?

**Yes.** This is the way. To wit, the prospective `beartype.abby` subpackage will export two companion utility functions:

* `beartype.abby.is_bearable()`, strictly returning a `bool`. This tester is guaranteed to:
  * *Never* raise an exception while type-checking.
  * *Only* raise an exception before type-checking if the passed type hint is either:
    * Invalid as a type hint, usually due to *not* being PEP-compliant.
    * Valid as a type hint but currently unsupported by `@beartype`. In theory, `@beartype` supports everything valid as a type hint. In practice, my promises are as good as smoke in the wind.
* `beartype.abby.die_if_unbearable()`, which does the exact opposite. Specifically, `die_if_unbearable()` *never* returns a `bool` but instead either:
  * If the passed object violates the passed type hint, raises some sort of a `BeartypeCallHintPepException` that frankly has yet to be determined. Someday we will all know more.
  * Else (i.e., if the passed object satisfies the passed type hint), silently does nothing. So, silence has a sound after all...

Too many words. Succinctly, `beartype.abby.is_bearable()` strictly returns booleans and only raises exceptions when passed "bad" type hints. Thus spake Leycecthustra.

**Comment by amacfie** [comment]
Why am I getting `is_bearable([2,3,4,"5"], list[int])  # ->> True`? 🤔

**Comment by TeamSpen210** [comment]
That's intentional. To allow for fast constant-time type checking, `beartype` doesn't exhaustively check the entirety of lists and other collections. Instead each time it's called, it randomly picks a position and checks it. Most of the time, you don't tend to have lists like your example with only a few bad items, it's usually entirely the wrong type so this will catch it immediately. If there are incorrect values, repeatedly calling will eventually produce a typecheck failure.

However this might not be a desirable behaviour for an explicit type-checking function, so it'd probably be better to default to the `O(n)` mode described here, once implemented: #60

**Comment by leycec** [comment]
> However this might not be a desirable behaviour for an explicit type-checking function...

You are not wrong. You never are. You know this already, which is why I'm generalizing the signature of our currently implemented and thoroughly untested `is_bearable()` tester to read:

``` python
def is_bearable(
    obj: object, hint: object,
    *, conf: BeartypeConf = BeartypeConf()
):
```

For orthogonality and (more importantly) *'cause this is @beartype*, `conf` will probably default to `BeartypeConf()` which defaults to our default `BeartypeStrategy.O1` strategy. Gotta type-check in constant time if you're gonna keep up with the big boys!

Now, let's see if I actually do something and manage to test any of this. The sweat pouring off my chin is real.
